### PR TITLE
Refactor "currently_fetching_*" flags

### PR DESCRIPTION
**Pull request by jerivas**

### CODE REVIEW

**Comment by jerivas** [comment]
[Prevent "currently_*" flags from getting stuck "on" if job/Heroku crashes.](https://trello.com/c/x9dD3PBe/298-prevent-currently-flags-from-getting-stuck-on-if-job-heroku-crashes)

File: metecho/api/models.py
Comment (jerivas) [inline metecho/api/models.py:140]: @jgerigmeyer @davisagli it turns out when refreshing user repos we didn't have any error handling in place. This makes the `currently_fetching_repos` flag get stuck if the job is interrupted. I have refactored the the job to be in line with how we handle other async tasks and this time I think we should get the flag set to `False` even if the worker is shut down.

I'll continue with the other changes listed on the card but just wanted to note this

File: metecho/api/models.py
Comment (jerivas) [inline metecho/api/models.py:123]: @jgerigmeyer Previously WS events were only being sent on success. This now emits events on error as well

**Review by jgerigmeyer — COMMENTED** [review]
Looks good! I added some front-end handling for the new `USER_REPOS_ERROR` event.

File: metecho/api/jobs.py
Comment (davisagli) [inline metecho/api/jobs.py:606]: Why are you explicitly logging the traceback before re-raising? The traceback for the raised exception should end up getting recorded in the failed rq job.

Comment (jerivas) [reply metecho/api/jobs.py:606]: I'm following the pattern from the other jobs, they all seem to do that. Perhaps the intention is to get a formatted traceback in the logs?

Comment (davisagli) [reply metecho/api/jobs.py:606]: I guess I have no problem with it if it's helpful. For myself, I'm always investigating after the fact when I have access to failed rq jobs but not old Heroku logs

File: metecho/api/models.py
Comment (jerivas) [inline metecho/api/models.py:383]: @jgerigmeyer I removed this call to the job because we `save()` when finalizing the job so this just re-queued the job again. I assume this was here so GitHub users are fetched in the background after a new project is added in the admin, but I think we can have the frontend queue the job as we do in other places when the required data is empty (in this case `project.github_users`). What do you think?

File: metecho/api/models.py
Comment (jgerigmeyer) [inline metecho/api/models.py:321]: We were already disabling the btn on the front end, but it wouldn't persist -- so this is probably better 👍

File: metecho/api/models.py
Comment (jerivas) [reply metecho/api/models.py:367]: I think that would be worth addressing now. My proposal would be to do it only during an admin save, or even adding a dedicated admin button to queue the image refresh

Comment (jgerigmeyer) [reply metecho/api/models.py:367]: I like that. @davisagli, does that sound okay to you?

Comment (davisagli) [reply metecho/api/models.py:367]: Yes please; admin save is the only time we actually want to try to refresh the image.

File: src/js/index.tsx
Comment (jerivas) [inline src/js/index.tsx:214]: @jgerigmeyer this seems to do what we want: disable the "Re-sync projects" button if the flag is stuck in the backend. Not sure if this was the right place to add it though

Comment (jgerigmeyer) [reply src/js/index.tsx:214]: @jerivas Practically, these two branches (`REFRESH_PROJECTS_REQUESTED` vs `projectsRefreshing()`) are equivalent. So the real change here is marking projects as "loading" even if there are already existing projects. The issue here is a race condition though, and this solves one problem but creates another.

When a user initially logs in, the front-end fetches the list of projects, and simultaneously the back-end [automatically triggers a `user.queue_refresh_repositories()`](https://github.com/oddbird/Metecho/blob/main/metecho/api/models.py#L1309-L1311). The problem is that this happens before the client has had a chance to subscribe to the relevant websocket channels, and often the job even _completes_ before the WS channel subscription is finalized. This means that the client usually does not receive the final `USER_REPOS_REFRESH` event, which would signal that the job is done and it's time to refetch the new list.

The prior "solution" to this (which was not really a solution) was to only show the loading indicator if there were no projects. _Usually_ this meant the following order of events:

1. User logs in.
2. Back-end triggers refresh.
3. Front-end loads user model, sees that the user is currently re-fetching projects.
4. Re-fetch repos job completes.
5. Front-end loads projects list.
6. Front-end subscribes to WS channels.

If (5) happens before (4) and the list was not empty, then the front end never knows that the job completed, and never re-fetches the updated list. If the list was empty, then the front end often hangs in the "loading" state until a page refresh (unless the WS subscription was able to happen before the job completed).

So this was already broken, in addition to the "what happens if the `user.currently_fetching_repos` flag is stuck on?" problem. But this change doesn't really fix it, but rather exposes the race condition issue to all users on every login (or almost every login, depending).

We need a better solution here, but I'm having trouble thinking of what it should be. If we can't come up with something reliable, maybe the front-end polls every second or so while `user.currently_fetching_repos` is on?

Comment (jerivas) [reply src/js/index.tsx:214]: @jgerigmeyer Interesting, didn't know so much was going on. Would it help to move this to the frontend explicitly requesting a project refresh? It seems we can get rid of some the uncertainty without that backend signal.

Comment (jgerigmeyer) [reply src/js/index.tsx:214]: > Would it help to move this to the frontend explicitly requesting the repos refresh? It seems we can get rid of some the uncertainty by removing that backend signal.

@jerivas Oh, that's an interesting proposal. It would add a bit of loading time on every login, but maybe that's okay. Or alternatively if we want to avoid that, we could skip the refresh unless the list is empty? I'll try it out and see how it feels.

Comment (jgerigmeyer) [reply src/js/index.tsx]: @jerivas In the end, explicitly requesting the refresh from the front-end only partially solved the issue, because there was still often a race condition where the WS subscription wasn't necessarily complete before the job finished. One option would be to wait until the WS subscription is made before even requesting the refresh or showing the app -- I played with that initially, but it felt a bit fragile to depend on WS before rendering.

Instead, I decided to just poll the API every second, to see if the job is done. If the job already completed before the WS subscription, then the front end will discover this after a 1 second delay. If the WS subscription was made before the job completes, then the WS event will come through normally and the polling will cancel. Hand-testing seems to confirm this approach is okay. What do you think about it?

https://github.com/oddbird/Metecho/commit/74dc92c86fd4762e04e3ac0a4cff01043efd0c72

**Review by jgerigmeyer — APPROVED** [review]
@jerivas I think this PR is good to go? Pending your review of my changes...

File: src/js/index.tsx
Comment (jerivas) [reply src/js/index.tsx]: I think it's a bit complex, but I don't see any other way to comply with what we need, so LGTM 👍

I would suggest setting a limit to the retries so we don't DDOS ourselves if something goes wrong.

Comment (jgerigmeyer) [reply src/js/index.tsx]: Agreed, I'll add a limit. I also was hoping for a simpler solution, but in the end this felt like the "correct" and more robust approach -- if we're relying on WS events but don't want to wait until WS subscriptions complete (they can be slow), then I feel like we need a fallback like this.

Comment (jgerigmeyer) [reply src/js/index.tsx]: @jerivas Limit added in https://github.com/oddbird/Metecho/commit/4ad24fa06872571f3199474dd1ef4f4a6ba92253. Go ahead and merge if this PR looks okay to you?

**Pull request merged**
